### ISSUE TITLE
The order of output does not match to order of input

### ISSUE THREAD

**Issue by NaSed** [Rdatatable/data.table#1171]
Hi,
I read a table with two columns. You can download it as RData file through the below link:
https://www.dropbox.com/s/sbijdvloz89oj0m/geneSymb_Entrez.RData?dl=0

After loading it in environment, I run the below instructions:
names_DT <- data.table(Symbol=names[,1], Entrez=names[,2])
names_DT[Entrez == c("2597","3945", "6222", "2597"),]
 and get result as:
   Symbol Entrez
1:  GAPDH   2597
2:  GAPDH   2597
3:   LDHB   3945
4:  RPS18   6222

As you can see the order of output is different from the input?! Why it happens? Replacing "==" with "%in%" does not make difference!
Am I missed something? 

Thanks.

**Comment by mattdowle** [comment]
Please try the latest version from GitHub before posting issues.   With v1.9.5,  I get:

```R
> DT = data.table(a=c(3,2,2,1),b=1:4)
> DT
   a b
1: 3 1
2: 2 2
3: 2 3
4: 1 4
> DT[a==c(2,1)]
Error in `[.data.table`(DT, a == c(2, 1)) : 
  RHS of == is length 2 which is not 1 or nrow (4). For robustness, no recycling is allowed (other than of length 1 RHS). Consider %in% instead.
> DT[a%in%c(2,1)]
   a b
1: 2 2  # original order is retained
2: 2 3
3: 1 4
> packageVersion("data.table")
[1] ‘1.9.5’
> 
```

Is this correct and desired behaviour now?   In this quick test it seems to me that %in% retains the original order.  If I've understood correctly, please confirm and close this issue.

3rd item in README : 

> DT[column == value] no longer recycles value except in the length 1 case (when it still uses DT's key or an automatic secondary key, as introduced in v1.9.4). If length(value)==length(column) then it works element-wise as standard in R. Otherwise, a length error is issued to avoid common user errors. DT[column %in% values] still uses DT's key (or an an automatic secondary key) as before. Automatic indexing (i.e., optimization of == and %in%) may still be turned off with options(datatable.auto.index=FALSE).

**Comment by DavidArenburg** [comment]
This is what I was practically trying to explain to the OP [here](http://stackoverflow.com/questions/30687428/changing-the-order-in-data-table-r) but I can't get thru for some reason...

**Comment by mattdowle** [comment]
@DavidArenburg  I read it and skimmed the comments.  Feel free to have the satisfaction of posting an answer something like "look folks, ALWAYS check the latest version before posting and discussing.  Actually install and try the latest version.  Actually search the README."

**Comment by NaSed** [comment]
Thank you @mattdowle .
Actually, my problem was that the code was not working for "my data", while for the other examples it was working.

This is an example:
names <- matrix(c("A","1","B","2","C","3","D","4"), ncol=2, byrow=T) 
names_DT <- data.table(Symbol=names[,1], Entrez=names[,2]) 
names_DT[Symbol==c("D","A","B","A"),]

Symbol Entrez
1:      D      4
2:      A      1
3:      B      2
4:      A      1

As you can see, I have defined the 'names' by myself and all of the other instructions are the same as what I have written in the comment; and the instructions work here. I mean the order of output matches with input and repeated element in input has been repeated in output, as well.
Two different responses for the same code and two different inputs data was my problem which had made me surprised!

Anyway, you're right. The defined manner in new version is more rational and I am trying to install version 1.9.5, now.
Thank you for your time and kindness.

**Comment by mattdowle** [comment]
I see.  Yes it was my mistake in v1.9.4 to try and make `%in%` and `==` equivalent.  Corrected in v1.9.5.  `==` works element-wise consistent with base R (and prior to v1.9.4) and `names_DT[Symbol==c("D","A","B","A"),]` now returns an empty data.table which is correct behaviour consistent with base (in this case because `names_DT` has 4 rows in this example).
